onnxruntime/core/providers/xnnpack/nn/conv.cc

```diff
@@ -149,6 +153,15 @@ ONNX_OPERATOR_VERSIONED_KERNEL_EX(Conv, kMSInternalNHWCDomain, 1, 10, kXnnpackEx
 ONNX_OPERATOR_KERNEL_EX(Conv, kMSInternalNHWCDomain, 11, kXnnpackExecutionProvider,
                         KernelDefBuilder().TypeConstraint("T", DataTypeImpl::GetTensorType<float>()),
                         Conv);
+#ifdef XNNPACK_FP16_SUPPORTED
```

Comment:
Not quite understand why do we have the macro here?
 Is this for reducing binary size or satisfying xnnpack's limitation?

Reply:
Yes. Good finding. I was mislead by KleidiAI.

Reply:
Good reminder.
According to 
```
OPTION(XNNPACK_ENABLE_AVX512FP16 "Build XNNPACK with AVX512-FP16 micro-kernels" ON)
IF(CMAKE_C_COMPILER_ID STREQUAL "GNU")
  IF(CMAKE_C_COMPILER_VERSION VERSION_LESS "13")
    SET(XNNPACK_ENABLE_AVX512FP16 OFF)
  ENDIF()
ELSEIF(CMAKE_C_COMPILER_ID STREQUAL "Clang")
  IF(CMAKE_C_COMPILER_VERSION VERSION_LESS "15")
    SET(XNNPACK_ENABLE_AVX512FP16 OFF)
  ENDIF()
ENDIF()
```
https://github.com/google/XNNPACK/blob/89c542702b1bb3eb9f995b973045e13490d3c297/CMakeLists.txt#L209C1-L218C8.
I think the macro is necessary.
And I should add AVX512FP16 for this macro.

Reply:
Got it. Thanks

Reply:
Given we're only enabling for ARM64 currently AVX512 isn't relevant is it?

Reply:
Yes. But I'm a little confused that it looks the onnxruntime_use_avx512 is disabled by default. And I don't find it's enabled in the repo.

Reply:
maybe I need to add a --use_avx512 in build.py to make it consistent.

Reply:
I want to update it in the next PR.

Reply:
I don't think onnxruntime_mlas.cmake looks at onnxruntime_use_avx512, so it's possible that setting does nothing.

Reply:
According to https://learn.microsoft.com/en-us/cpp/build/reference/arch-x64?view=msvc-170, /arch:AVX512 is necessary for VC.
Maybe we don't set it on ORT side because xnnpack has set it
SET_PROPERTY(SOURCE ${ALL_F16C_MICROKERNEL_SRCS} APPEND_STRING PROPERTY COMPILE_FLAGS " /arch:AVX ")
https://github.com/google/XNNPACK/blob/14ce5c47ebf2ff79e3dc2f1d319ad7b1793e1b21/CMakeLists.txt#L868C5-L868C105
I'll do some tests on windows

Reply:
Actually, To detect avx512/avx2/avx/sse is a runtime behavior.

https://github.com/microsoft/onnxruntime/blob/66fbf7df902ad889198834c83251db848f647adf/onnxruntime/core/mlas/lib/platform.cpp#L434